src/datasets/arrow_dataset.py

```diff
@@ -2175,6 +2178,50 @@ def generator():
         logger.info(f"Finished writing TFRecord to {filename}")
         self = None  # delete the dataset reference used by tf_dataset
 
+    def _write_csv(self, file_obj: BinaryIO, batch_size: int, **to_csv_kwargs) -> int:
+        """
+        Writes the pyarrow table as CSV to a binary file handle.
+        Caller is responsible for opening and closing the handle.
+        """
+        written = 0
+        header = to_csv_kwargs.pop("header", True)
+        encoding = to_csv_kwargs.pop("encoding", "utf-8")
+        to_csv_kwargs.pop("path_or_buf", None)
+
+        for offset in range(0, self._data.num_rows, batch_size):
+            batch = self._data.slice(offset, offset + batch_size)
```

Comment:
```suggestion
            batch = query_table(
                self._data, slice(offset, offset + batch_size), indices=self._indices.column(0) if self._indices is not None else None
            )
```
This should add support for shuffled dataset as well (or any dataset after a shuffle/shard/select transform)
`self._indices` is the mapping between the indices of the dataset the indices of the table.